### PR TITLE
✨ Add support for managed identity in machine pool

### LINKED ISSUE #960: Add support for managed identity with AzureMachinePool
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/312 and https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/616 added support for Azure Managed identities for AzureMachines. We should port the changes to AzureMachinePool to allow VMSS using MSI.

This will become especially useful if/when we add support for VMSS control planes.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by shysank**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds support for managed identity in machine pool. 
- Ports #616 changes to machine pool

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #960 

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release Note**:

```release-note
Add support for managed identity in machine pool
```

### CODE REVIEW

File: cloud/converters/identity.go
Comment (shysank) [inline cloud/converters/identity.go:47]: Too many 'S's here?

**Review by nader-ziada — COMMENTED** [review]
Thanks for the PR @shysank 

a couple of quick questions

File: cloud/services/virtualmachines/virtualmachines.go
Comment (shysank) [reply cloud/services/virtualmachines/virtualmachines.go:159]: I refactored this, and moved it to converters [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/47d4027e55b281e23b62aa7a87df8dabad479727/cloud/converters/identity.go#L33). Since the logic for vmss is identical to vm, I thought it would be easier to read. Let me know, if it's okay.

File: cloud/converters/identity.go
Comment (shysank) [reply cloud/converters/identity.go:28]: Yes, this is just a helper for converting `[]UserIdentity` to `map[string]VmOrVmssIdentityStruct`

File: cloud/services/virtualmachines/virtualmachines.go
Comment (nader-ziada) [reply cloud/services/virtualmachines/virtualmachines.go:159]: oh ok, got it, sounds good, I will run this template a bit and test it out before putting in more comments

**Comment by nader-ziada** [comment]
@devigned can you take a look please

**Comment by nader-ziada** [comment]
@shysank if you think the PR is ready, can you please remove the [WIP] and add a release note in the PR description.

File: templates/flavors/machinepool-user-assigned-identity/patches/user-assigned-identity.yaml
Comment (devigned) [inline templates/flavors/machinepool-user-assigned-identity/patches/user-assigned-identity.yaml:9]: newline, please

File: templates/flavors/machinepool-user-assigned-identity/kustomization.yaml
Comment (devigned) [inline templates/flavors/machinepool-user-assigned-identity/kustomization.yaml:5]: newline, please

File: templates/flavors/machinepool-system-assigned-identity/patches/system-assigned-identity.yaml
Comment (devigned) [inline templates/flavors/machinepool-system-assigned-identity/patches/system-assigned-identity.yaml:7]: newline, please

File: templates/flavors/machinepool-system-assigned-identity/kustomization.yaml
Comment (devigned) [inline templates/flavors/machinepool-system-assigned-identity/kustomization.yaml:5]: newline, please

File: config/crd/bases/exp.infrastructure.cluster.x-k8s.io_azuremachinepools.yaml
Comment (devigned) [inline config/crd/bases/exp.infrastructure.cluster.x-k8s.io_azuremachinepools.yaml:316]: I think I know what this means; the user is responsible for making the user assigned identity in Azure on their own. If this is the case, it would be great to link to some documentation about creating a user assigned identity. In other words, can be more specific and action oriented to help a user to figure this out?

File: cloud/services/scalesets/vmss_test.go
Comment (devigned) [inline cloud/services/scalesets/vmss_test.go:721]: *No action in the PR, but more for maintainers.*

Dang... these tests are verbose with only like 2 lines that need to be updated. It would be nice to dry these up.

cc: @CecileRobertMichon and @nader-ziada

File: cloud/converters/identity.go
Comment (devigned) [inline cloud/converters/identity.go:63]: `strings.HasPrefix(..)` is superfluous.

https://play.golang.org/p/a9TuyaEINjP

File: cloud/converters/identity_test.go
Comment (devigned) [inline cloud/converters/identity_test.go:23]: please sort the imports

File: cloud/converters/identity_test.go
Comment (devigned) [inline cloud/converters/identity_test.go:26]: this doesn't look formatted

**Review by devigned — CHANGES_REQUESTED** [review]
Looks pretty good. I have some light feedback about the code and documentation.

One thing I would like to see out of this PR is a tutorial in the CAPZ book to describe how to create a MachinePool with a user assigned identity with a link to building a user assigned identity.

**Comment by CecileRobertMichon** [comment]
re: documentation, you could add onto the existing Identity doc https://capz.sigs.k8s.io/topics/identity.html

**Comment by shysank** [comment]
/retitle ✨ Add support for managed identity in machine pool

**Comment by shysank** [comment]
@devigned I've addressed the review comments, PTAL. 
cc: @CecileRobertMichon @nader-ziada

**Review by devigned — APPROVED** [review]
lgtm

/assign @nader-ziada @CecileRobertMichon

**Comment by nader-ziada** [comment]
lgtm pending squashing commits

File: cloud/services/scalesets/vmss.go
Comment (CecileRobertMichon) [inline cloud/services/scalesets/vmss.go:199]: for system assigned, where is the role assignment being reconciled for AzureMachinePools? similar to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/controllers/azuremachine_reconciler.go#L82 for AzureMachines

Comment (devigned) [reply cloud/services/scalesets/vmss.go:199]: I didn't call this out b/c I was assuming the identity has been created with the roles needed before being used. I think this indicates that we should be specific about the expectations here.

Comment (CecileRobertMichon) [reply cloud/services/scalesets/vmss.go:199]: That's for user assigned. For system assigned the identity is tied to the VM's lifecycle so it can't be created ahead of time by the user no?

Comment (devigned) [reply cloud/services/scalesets/vmss.go:199]: Oops! You are totally right.

Comment (shysank) [reply cloud/services/scalesets/vmss.go:199]: Thanks @CecileRobertMichon , I think I missed the role assignment part for system assigned identity. I'll take a look into it.

**Comment by CecileRobertMichon** [comment]
/retest

looks like an unrelated flake

**Comment by shysank** [comment]
@CecileRobertMichon I have made the role assignment changes for system assigned identity, PTAL. The api-diff script is failing because I modified the `Validate` method to accept old machine pool spec as a parameter.

**Comment by nader-ziada** [comment]
I think its lgtm pending squashing commits

**Review by CecileRobertMichon — APPROVED** [review]
lgtm, please squash the commits @shysank 

Awesome work, thanks!

**Comment by shysank** [comment]
@nader-ziada @CecileRobertMichon Thanks for the review, I've squashed the commits.

**Review by devigned — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
